### PR TITLE
Link related papers

### PR DESCRIPTION
**Pull request by Joozty**
@viktor-yakubiv  I decided to add that floating notification we were talking about to the bottom toolbar after all. I didn't like it when it was in the right top corner. It was kind of disturbing. Also, I implemented back to the reading button in the toolbar. Please have a look 🙂

### CODE REVIEW

File: polyfills.js [hunk 1]
```diff
@@ -1,4 +1,8 @@
 // Support for IE
 // https://stackoverflow.com/a/53332776/5594539
+import smoothscroll from 'smoothscroll-polyfill'
```
Comment (viktor-yakubiv) [inline polyfills.js]: This post saids that you do not need anything because `NodeList.prototype.forEach` is polyfilled by core.js. However, there is nothing about smooth scroll.

Previously we agreed that we do not really care about IE. This polyfill adds behaviour to Edge and Safari but this is not a big deal too.

I would try to avoid CSS polyfills unless it's really necessary but this is up to you. Just keep in mind that adding any extra JS harms user's performance. Not everybody uses machines with high level performance as like as we do.

File: reader/components/pdf-outline-sidebar/PDFOutlineSidebar.jsx [hunk 2]
```diff
@@ -30,13 +31,26 @@ class PDFOutlineSidebar extends React.PureComponent {
...
 
-    if (!outline || outline.length === 0)
-      return <h5>PDF does not contain any outline</h5>
+    const finalOutline = [
+      ...(outline || []),
+      {
+        onClick: () => {
+          dispatch(scrollToRelatedPapers())
+        },
+        title: 'Related papers',
```
Comment (viktor-yakubiv) [inline reader/components/pdf-outline-sidebar/PDFOutlineSidebar.jsx]: Can we rename the section to 'Related papers' too?

File: reader/components/pdf-outline-sidebar/renderers/OutlineRenderer.jsx [hunk 3]
```diff
@@ -3,22 +3,24 @@ import { Collapse } from 'reactstrap'
...
-      </ol>
-    </Collapse>
+    <>
+      <Collapse isOpen={isExpanded}>
+        <ol>
+          {outline.map((item, index) => {
+            /* eslint-disable react/no-array-index-key */
+            return (
+              <OutlineItem
+                key={`${index}${isExpanded}`}
```
Comment (viktor-yakubiv) [inline reader/components/pdf-outline-sidebar/renderers/OutlineRenderer.jsx]: If I understand React's reconciliation, it's better to have bare index only.

File: reader/components/pdf-recommender/PDFRecommender.jsx [hunk 4]
```diff
@@ -1,10 +1,38 @@
...
+        () =>
+          recommenderRef.current.parentNode.scroll({
+            left: 0,
+            top: offsetTop - 200,
+            behavior: 'smooth',
+          }),
+        50
+      )
+    }
+  }, [ui.isRelatedPapersScrolled])
```
Comment (viktor-yakubiv) [inline reader/components/pdf-recommender/PDFRecommender.jsx]: Seems to be really interesting 🙂

Have you tested this on small papers and small windows? (lower than 1000 pixels, and lower than 200 pixels)

File: reader/components/pdf-toolbar/PDFToolbar.jsx [hunk 5]
```diff
@@ -29,20 +38,78 @@ const PDFToolbar = ({ viewer, eventBus }) => {
...
+      <div className="pdf-related-papers d-flex flex-row align-items-center justify-content-between mr-lg-5 mr-2 order-3">
+        <button
+          title="Show related papers"
+          type="button"
+          className="btn m-auto"
+          onClick={() => setRelatedPapersClicked(!relatedPapersClicked)}
+        >
+          {!relatedPapersClicked ? 'Related papers' : 'Back to reading'}
+        </button>
+      </div>
```
Comment (viktor-yakubiv) [inline reader/components/pdf-toolbar/PDFToolbar.jsx]: I would like to slightly change this but I am not sure in which way. Basically, I like your solution, it's quite creative but I am not sure I would like to have this button as a part of the interface.

I suggest to fix the padding of the button and keep it as it is.

File: reader/components/main-area/MainArea.scss [hunk 6]
```diff
@@ -38,7 +38,9 @@ $sidebar-width: 20rem;
   transition: left 200ms ease;
   will-change: left;
 
-  .sidebar-open & {
-    left: $sidebar-width;
+  @include media-breakpoint-up(md) {
+    .sidebar-open & {
+      left: $sidebar-width;
```
Comment (viktor-yakubiv) [inline reader/components/main-area/MainArea.scss]: I feel like this is not related to the links? This is the sidebar keeping in the screen above `md`. Is that correct?

File: reader/components/pdf-toolbar/PDFToolbar.jsx [hunk 7]
```diff
@@ -1,15 +1,24 @@
...
 
 const MIN_SCALE = 0.1
 const MAX_SCALE = 10.0
 const DEFAULT_SCALE_DELTA = 1.1
 
 const PDFToolbar = ({ viewer, eventBus }) => {
+  const [{ ui }, dispatch] = useGlobalStore()
+  const [readPosition, setReadPosition] = useState(0)
+  const [relatedPapersClicked, setRelatedPapersClicked] = useState(null)
+  const [isVisible, changeVisibility] = useState(false)
```
Comment (viktor-yakubiv) [inline reader/components/pdf-toolbar/PDFToolbar.jsx]: I would prefer complex state object here instead the list of `useState` calls. It's just easier to read and understand what the state is about even if it's more lines of code.

Rect hooks are introduces to reduce complexity in cases where it's too much to have a class. However, there are cases where functional approach does not fit as well as object oriented does.

File: reader/components/main-area/MainArea.scss (hunk 6 above)
Comment (Joozty) [reply reader/components/main-area/MainArea.scss]: Yes, it is. Maybe I could have created separate PR.

File: reader/components/pdf-toolbar/PDFToolbar.jsx (hunk 7 above)
Comment (Joozty) [reply reader/components/pdf-toolbar/PDFToolbar.jsx]: I do agree. It doesn't have to be written as class. Instead of using `useState` several times I could have used one useReducer. I was just too lazy to rewrite it.

File: reader/components/pdf-recommender/PDFRecommender.jsx (hunk 4 above)
Comment (Joozty) [reply reader/components/pdf-recommender/PDFRecommender.jsx]: Not really. Will do it. Thanks

File: polyfills.js (hunk 1 above)
Comment (Joozty) [reply polyfills.js]: The StackOverflow comment is for code bellow. (I will move it closer to relevant code.) Nonetheless, it seems polyfill `forEach` is not needed. I experienced troubles with it in the first stages of Reader. Apparently it has been fixed in new versions of `core-js` or `pdfjs`.


Smooth scroll polyfill is needed otherwise app fails when `related papers` link is clicked. It could be handled with `try/except block` but I like to have smooth scroll there anyway.

```
The above error occurred in the <PDFRecommender> component:
    in PDFRecommender (created by PDFViewer)
    in div (created by PDFViewer)
    in PDFViewer (created by MainArea)
    in Unknown (created by MainArea)
    in div (created by MainArea)
    in MainArea (created by CoreReader)
    in div (created by Layout)
    in Layout (created by CoreReader)
    in GlobalProvider (created by CoreReader)
    in CoreReader (created by ForwardRef(LoadableComponent))
    in ForwardRef(LoadableComponent) (created by Reader)
    in Reader (created by WithAnalytics)
    in WithAnalytics (created by App)
    in App
    in Container (created by AppContainer)
    in AppContainer

React will try to recreate this component tree from scratch using the error boundary you provided, Reader.
dll_01ec57fc9b90d43b98a8.js (22151,5)

SCRIPT438: SCRIPT438: Object doesn't support property or method 'scroll'
```

File: reader/components/pdf-recommender/PDFRecommender.jsx (hunk 4 above)
Comment (Joozty) [reply reader/components/pdf-recommender/PDFRecommender.jsx]: It works just fine at least what I tested.

File: reader/components/pdf-outline-sidebar/renderers/OutlineRenderer.jsx (hunk 3 above)
Comment (Joozty) [reply reader/components/pdf-outline-sidebar/renderers/OutlineRenderer.jsx]: Can you point me to some articles?

File: reader/components/pdf-outline-sidebar/PDFOutlineSidebar.jsx (hunk 2 above)
Comment (Joozty) [reply reader/components/pdf-outline-sidebar/PDFOutlineSidebar.jsx]: https://github.com/oacore/reader/pull/89/commits/3b2a94373692b1d658ed28592882a252b0943a27

File: reader/components/pdf-toolbar/PDFToolbar.jsx (hunk 5 above)
Comment (Joozty) [reply reader/components/pdf-toolbar/PDFToolbar.jsx]: https://github.com/oacore/reader/pull/89/commits/bf8d8a6d02c3004eda34f43ab864585a55dfdf99 I believe you meant it this way and not changing the height of the whole element (it looked ugly).

File: polyfills.js (hunk 1 above)
Comment (Joozty) [reply polyfills.js]: https://github.com/oacore/reader/pull/89/commits/68ba3bc7fc69facc938047c4453ca405af11d6d3

File: reader/components/pdf-outline-sidebar/renderers/OutlineRenderer.jsx [hunk 8]
```diff
@@ -3,22 +3,24 @@ import { Collapse } from 'reactstrap'
...
-              item={item}
-              linkService={linkService}
-              isExpanded={isExpanded}
-            />
-          )
-          /* eslint-enable */
-        })}
-      </ol>
-    </Collapse>
+    <>
```
[the remark this answers is not in the document]
Comment (Joozty) [reply reader/components/pdf-outline-sidebar/renderers/OutlineRenderer.jsx]: https://github.com/oacore/reader/pull/89/commits/48d7a5a88a10bd8f2ae3e1cadbe79c24d88da86f

File: reader/components/pdf-toolbar/PDFToolbar.jsx (hunk 7 above)
Comment (viktor-yakubiv) [reply reader/components/pdf-toolbar/PDFToolbar.jsx]: That's not really good in general. However, it's okay to be lazy 🙂

File: reader/components/pdf-toolbar/PDFToolbar.jsx (hunk 5 above)
Comment (viktor-yakubiv) [reply reader/components/pdf-toolbar/PDFToolbar.jsx]: Yes, that is what I meant. Thank you

File: reader/components/pdf-recommender/PDFRecommender.jsx (hunk 4 above)
Comment (viktor-yakubiv) [reply reader/components/pdf-recommender/PDFRecommender.jsx]: Perhaps API works around negative numbers somehow in a good way. It is expectable if think a little bit about but we had to test it.

Thank you!

Comment (Joozty) [reply reader/components/pdf-recommender/PDFRecommender.jsx]: Btw. I am working on some basic Cypress tests. I will create PR once Github actions are released.

File: reader/components/pdf-outline-sidebar/renderers/OutlineRenderer.jsx (hunk 3 above)
Comment (viktor-yakubiv) [reply reader/components/pdf-outline-sidebar/renderers/OutlineRenderer.jsx]: Explanation of [Reconciliation in the React documentation](https://reactjs.org/docs/reconciliation.html) might be the best I can send you.

The idea of key is to help React to understand that equal elements, i.e. types are equal, are different and vice versa. In other words, React will create new node for the elements with new keys and remove nodes for elements that keys has disappeared.

Coming back to your case, when the item becomes expanded/collapsed, React removes the item HTML element from the tree, creates and places a new one. It might be somehow optimised, e.g. elements may be reused but I cannot be sure because I have not explored the code.

Comment (Joozty) [reply reader/components/pdf-outline-sidebar/renderers/OutlineRenderer.jsx]: Make sense. I think the reason for using ```key{`${index}${isExpanded}`}``` was to get rid of `eslint-disable`. But obviously it didn't work. 😄
